problem_builder/mixins.py

```diff
@@ -0,0 +1,123 @@
+
+
+class EnumerableChildMixin(XBlockWithTranslationServiceMixin):
+    CAPTION = _(u"Child")
+
+    show_title = Boolean(
+        display_name=_("Show title"),
+        help=_("Display the title?"),
+        default=True,
+        scope=Scope.content
+    )
+
+    @lazy
+    def siblings(self):
+        return self.get_parent().steps
+
+    @lazy
+    def step_number(self):
+        return list(self.siblings).index(_normalize_id(self.scope_ids.usage_id)) + 1
+
+    @lazy
+    def lonely_step(self):
+        if _normalize_id(self.scope_ids.usage_id) not in self.siblings:
+            message = u"{step_caption}'s parent should contain {step_caption}".format(step_caption=self.CAPTION)
+            raise ValueError(message, self, self.siblings)
+        return len(self.siblings) == 1
+
+    @property
+    def display_name_with_default(self):
+        """ Get the title/display_name of this question. """
+        if self.display_name:
+            return self.display_name
+        if not self.lonely_step:
+            return self._(u"{step_caption} {number}").format(
+                step_caption=self.CAPTION, number=self.step_number
+            )
+        return self._(self.CAPTION)
+
+
+class StepParentMixin(object):
+    """
+    An XBlock mixin for a parent block containing Step children
+    """
+
+    @property
+    def steps(self):
+        """
+        Get the usage_ids of all of this XBlock's children that are "Steps"
+        """
+        return [
+            _normalize_id(child_id) for child_id in self.children if child_isinstance(self, child_id, QuestionMixin)
+        ]
+
+
+class QuestionMixin(EnumerableChildMixin):
+    """
+    An XBlock mixin for a child block that is a "Step".
+
+    A step is a question that the user can answer (as opposed to a read-only child).
+    """
+    CAPTION = _(u"Question")
+
+    has_author_view = True
+
+    # Fields:
+    display_name = String(
+        display_name=_("Question title"),
+        help=_('Leave blank to use the default ("Question 1", "Question 2", etc.)'),
+        default="",  # Blank will use 'Question x' - see display_name_with_default
+        scope=Scope.content
+    )
+
+    def author_view(self, context):
+        context = context.copy() if context else {}
+        context['hide_header'] = True
+        return self.mentoring_view(context)
+
+    def author_preview_view(self, context):
+        context = context.copy() if context else {}
+        context['hide_header'] = True
+        return self.student_view(context)
+
+    def assessment_step_view(self, context=None):
+        """
+        assessment_step_view is the same as mentoring_view, except its DIV will have a different
+        class (.xblock-v1-assessment_step_view) that we use for assessments to hide all the
+        steps with CSS and to detect which children of mentoring are "Steps" and which are just
+        decorative elements/instructions.
+        """
+        return self.mentoring_view(context)
```

Comment:
@itsjeyd current implementation of `siblings` won't work for questions; they should provide their own (and as I said already, `EnumerableChildMixin` should not provide default implementation for `siblings`)